File: lib/code_gen/translator.rb
# typed: strict
# frozen_string_literal: true

require "sorbet-runtime"
require_relative "code_gen"
require_relative "../il"

module Lilac
  module CodeGen
    class Translator
      extend T::Sig
      extend T::Helpers

      abstract!

      include CodeGen

      sig do
        params(transformer: ILTransformer, cfg_program: IL::CFGProgram).void
      end
      def initialize(transformer, cfg_program)
        @transformer = transformer
        @program = cfg_program
      end

      sig { abstract.returns(Component) }
      def translate; end
    end
  end
end

File: Lilac/src/CodeGen/Translator.cs
using Lilac.Analysis;

namespace Lilac.CodeGen;

public abstract class Translator<TInstr> where TInstr : Instruction {
  protected ILTransformer<TInstr> Transformer { get; set; } = null!;
  protected CFGProgram CFGProgram { get; set; } = null!;

  public Translator(ILTransformer<TInstr> transformer,
                    CFGProgram cfgProgram) {
    Transformer = transformer;
    CFGProgram = cfgProgram;
  }

  // NOTE: for use by Translators that cannot use the default constructor.
  protected Translator() {
    // Empty
  }

  public abstract Component Translate();
}